Vertical funnel breakdowns

alexkim205 opened the pull request:
## Changes

Implementation of @clarkus's mockups in [#103](https://github.com/PostHog/product-internal/issues/103). Note that this PR significantly changes how our funnel step viz's look by combining the step bars with the table.

- [x] Ability to toggle breakdowns in table
- [x] Percentages of counts in y-axis
- [x] Edge cases in math used to construct bars
- [ ] Toggle to display percentages/absolute values (Punting to separate PR)
- [ ] Option for conversion rates visualization (Punting to separate PR)
- [x] Bar on hover and click
- [x] Table as legend

Note this branch merges in changes from #5730 because I'm anticipating changes to this viz will require the full window width. Merging this PR is blocked on #5730. Please refer to mockups in  [#103](https://github.com/PostHog/product-internal/issues/103) for more information.

Closes [#103](https://github.com/PostHog/product-internal/issues/103). 
Closes #5428.

**Before**


https://user-images.githubusercontent.com/13460330/133555567-03a0eb4c-a40f-4702-9120-f6b738335d8b.mov



**Adding/removing steps and different breakdowns**

https://user-images.githubusercontent.com/13460330/133554892-853780be-509b-48e8-92d0-eb811e42d250.mov

**Opening persons modal in different steps (+ baseline) and toggling legend**

https://user-images.githubusercontent.com/13460330/133554089-a89e6ea2-d3ff-4f2b-a53a-64242049ad04.mov

**Smoke regression test - vertical and horizontal layouts x w/ and w/o breakdowns**

https://user-images.githubusercontent.com/13460330/133555036-c933ef1f-f8e4-4c74-9707-469ca86e43bd.mov

**Infamous $current_url breakdown**


https://user-images.githubusercontent.com/13460330/133560730-861e2c4c-3c4b-48a4-ab9e-cee8c027f1e4.mov



## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

alexkim205 committed:
make funnels horizontal

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/funnels-horizontal-ui-proposal

alexkim205 committed:
move exclusion filter to horizontal tab

alexkim205 committed:
barebones implementation of vertical breakdowns

clarkus commented:
For the ghost bars, I think we can drop the patterned background for now. I was going to reserve those for comparison ranges at some point in the future. 

<img width="515" alt="Screen Shot 2021-08-25 at 9 40 12 AM" src="https://user-images.githubusercontent.com/254612/130830830-c617e0f3-bd5c-4a3e-a84b-3a5ce4650749.png">

In the examples above, is the white background around the third step's bars intentional?

alexkim205 commented:
> In the examples above, is the white background around the third step's bars intentional?

Good call, I had the contrast on my monitor setup so high that I didn't even notice this white bar. :'(

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/funnels-breakdown-vertical

alexkim205 committed:
refactor bar graph stuff a bit

paolodamico reviewed (commented):
Product-wise looking good so far! Tag me for a full review when ready 😉

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/funnels-breakdown-vertical

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/funnels-breakdown-vertical

alexkim205 committed:
add visibility toggle to table

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/funnels-breakdown-vertical

alexkim205 committed:
tie in table logic visibility map to bar display

clarkus commented:
+1 on giving users more control here. We need to identify upper and lower bounds for the count of bars in each step. There is going to be some upper bound where the visualization breaks down. I was working on this, but wound up moving on to other hot topics. I can revisit this week. 

Other considerations might include showing baseline conversion. There are cases where a user might want to compare a single value in a breakdown group to the baseline conversion of the funnel.

Twixes commented:
Will wait for Alex to handle next week.

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/funnels-breakdown-vertical

alexkim205 committed:
remove dummy data

alexkim205 committed:
add table legend

alexkim205 commented:
Update - adding table as a legend in this PR

![Screen Shot 2021-09-14 at 7 46 21 PM](https://user-images.githubusercontent.com/13460330/133362221-d5901dfe-8943-4021-9d60-01d4bc67e958.png)

alexkim205 commented:
There's still a few rough edges to work out but the new visualization functionally works

<img width="945" alt="Screen Shot 2021-09-15 at 4 00 49 PM" src="https://user-images.githubusercontent.com/13460330/133523985-0db28f9d-a41b-4aac-9689-c17f9021bc60.png">

alexkim205 committed:
add graphs to table

alexkim205 committed:
breakdown indices bug

alexkim205 committed:
fix toggling

alexkim205 committed:
fix regression tests and add bar interaction

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/funnels-breakdown-vertical

alexkim205 committed:
remove consoles

alexkim205 commented:
Finally ready for review! Tagging @mariusandra @paolodamico @clarkus for a 👀

alexkim205 committed:
edge cases with long steps and many breakdowns

marcushyett-ph commented:
Wow this is looking epic @alexkim205 - I'm curious as to how we're going to annotate the bars (I see both no annotation and numeric - but not sure if that makes sense if we're rendering %age)?

mariusandra reviewed (dismissed):
1. This seems really cool, but also like a pretty big difference. There's no way I'm able to review that all the numbers in the table/chart/etc are perfect with every combination of data. Plus it might also be a subjective change that should get a bit more feedback. 

Thus, an annoying request: can you add all the new changes behind a feature flag? Feel free to use new features flags view in the toolbar for magic :).

2. The visibility of the bars is off by default. They should always show.

![2021-09-16 17 49 27](https://user-images.githubusercontent.com/53387/133644444-f66ca2ea-a7a6-41c3-a7b6-17d51dc34898.gif)

3. Introducing yet another local state (local = in kea, but not synced to the URL/filters) for this visibility seems like introducing tech debt though. Any way to hook it up to the URL... and before making an API request to see if it's needed?

4. If I breakdown by something with just 1 prop, I'd expect these lines to be different. Isn't "baseline" just null?

![image](https://user-images.githubusercontent.com/53387/133645153-1e656b6a-2599-4038-af5b-427eafc5445e.png)

5. I kind of expect nothing to show in the funnel after this:

![2021-09-16 17 57 25](https://user-images.githubusercontent.com/53387/133645529-090a19ba-fd9e-41d9-85ac-f8a05e6c12b1.gif)

(not sure if this is scope creep?)

alexkim205 commented:
> Wow this is looking epic @alexkim205 - I'm curious as to how we're going to annotate the bars (I see both no annotation and numeric - but not sure if that makes sense if we're rendering %age)?

Thanks! Right now numeric counts will be shown above each bar if we're looking at 5 or less breakdown bars at once and if there's any more, I have it so that labels are hidden right now. I think it was brought up previously that having the numbers and percentages in the table align with each step would be a good enough visual indicator for dropoff. I feel like having extra annotation might get redundant, but maybe the table data isn't clear enough or should be highlighted better?

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/funnels-breakdown-vertical

marcushyett-ph commented:
All clear to me - I was just curious the rationale makes a lot of sense

alexkim205 commented:
Thanks Marius for the thorough review :)

1. FF's in toolbar makes it a lot easier to do that! I'll hide this behind a FF
2. fixed
3. stored state in URL instead of separate kea value
4. From my understanding and how I implemented it, baseline is supposed to represent the total of the breakdown bars. If there's only one breakdown property however, I think it makes sense to hide baseline. 
5. fixed

alexkim205 committed:
fixes from review

alexkim205 committed:
hide behind FF

mariusandra commented:
Looks cool, and it's definitely fun to enable/disable the feature with the toolbar. Some more things:

1. With the flag off:

![image](https://user-images.githubusercontent.com/53387/133670272-c9ddade1-9808-42dd-935c-db966db0116b.png)

... it still looks different than `master`:

![image](https://user-images.githubusercontent.com/53387/133670525-f7c2ffab-3e64-4694-8bcc-955ba454077a.png)

2. The visibility filter itself it broken:

![2021-09-16 21 11 10](https://user-images.githubusercontent.com/53387/133671377-b61895ca-fda5-42d0-9283-ea2c5ac139a1.gif)

My suggestions:
- instead of a visibility map, convert it to a hidden map. This way everything is visible by default, and `visible == !hidden[key]`, instead of `visible == visible[key] === undefined || visible[key]`.
- Index the visibility by key, and let's define a schema or it'll get out of hand quickly. If you index by index and the breakdown order changes from one day to the next, it'll fail. 
- Regarding the schema, something like `hidden = { "event/$pageview/${breakdownValue}": true, "action/1": true }` is probably enough. 
- Keep in mind this will also be used in the future to save the state of the retention toggles, and who knows what else.

clarkus reviewed (approved):
These changes look good to me overall. The only thing I think we should consider changing is the label for the funnel layout control. Top-to-bottom and Left-to-right aren't very descriptive of the layouts now. Top to bottom is visualizing breakdowns using stacked bars, but left to right is using distinct bars. I'm trying to think of alternate labels, but coming up blank. Probably fine to ship with these labels, but I feel like we could come up with something better. I will keep thinking about it and file a follow up issue.

paolodamico commented:
Looks like @mariusandra has reviewed the code and @clarkus the UI/UX, so withdrawing from review to avoid duplicated effort.

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/funnels-breakdown-vertical

alexkim205 committed:
actually hide behind FF, fix time conversion bug, and convert visibility map to hidden map

alexkim205 commented:
The hidden map concept is pretty sleek, thanks @mariusandra. Recent changes include: 

- Pushed up fixes to use a hidden map instead
- refactored and standardized visibility indices
- fixed a time conversion bug
- few more styling bugs that broke w/ merge conflicts

mariusandra commented on frontend/src/scenes/funnels/funnelLogic.ts:
```diff
@@ -80,6 +83,7 @@ export const cleanFunnelParams = (filters: Partial<FilterType>, discardFiltersNo
             : {}),
         ...(filters.funnel_window_interval ? { funnel_window_interval: filters.funnel_window_interval } : {}),
         ...(filters.funnel_order_type ? { funnel_order_type: filters.funnel_order_type } : {}),
+        ...(filters.hidden_map ? { hidden_map: filters.hidden_map } : {}),
```
Nit that won't hold up the PR: could we find a name that better describes what this does? The `map` here doesn't really say what is hidden. Even `hiddenLines` would be more descriptive, if not slightly misleading...

alexkim205 replied:
Hm, would `hiddenLegendKeys` work in this context?

mariusandra commented on frontend/src/scenes/funnels/funnelLogic.ts:
```diff
@@ -566,6 +673,16 @@ export const funnelLogic = kea<funnelLogicType>({
+                    ?.filter((s) => !!s.breakdown)
+                    ?.slice(5)
+                    .forEach((b) => {
+                        actions.setHiddenById({ [getVisibilityIndex(step, b.breakdown_value)]: true })
```
The comment is slightly misleading, as this sets the visibility of everything except the first 5?

alexkim205 replied:
got it! will update the comment

mariusandra reviewed (approved):
It's feature flagged, so let's get it in and start testing!

Things to fix in a follow up PR:
- Small nit in the code about the hidden map
- This seems broken, possibly thanks to the hidden map.
![2021-09-21 13 50 35](https://user-images.githubusercontent.com/53387/134165521-a91fbb0f-0f04-4006-aa01-bf66b55d0feb.gif)

mariusandra committed:
Merge branch 'master' into feat/funnels-breakdown-vertical

alexkim205 commented:
Follow up PR here #6061.
